### PR TITLE
Refactor map::add_item_or_charges

### PR DESCRIPTION
**Pull request by mugling**
Some of our item handling code has both poor documentation and performance. This PR starts an overhaul starting with a refactor of `map::add_item_or_charges`
### Performance improvements
- We now pass `const item &` avoiding a costly copy-construction for each call
- Check target tile for validity once (previously twice)
### Consistency
- 2D overload returns `item &` consistent with the 3D function (previously `bool`)
### Fixes inconsistent overflow radius

Previously we used various arguments including `1`, `2` and occasionally implicit conversions from `bool`. Now map placement either overflows to adjacent tiles or doesn't. There are only three occasions when it doesn't. Two of these are in the maple tree code and although potentially unintentional (ping @remyroy?) are left unchanged here.
### Documentation
- Update flag documentation (`NO_DROP`, `NOITEM`, `DESTROY_ITEM`)
- Update function `doxygen` comments
- Split conditionals in function to documented lambdas

### CODE REVIEW

**Comment by codemime** [comment]
Passing boolean flags to functions isn't particularly nice. Boolean flags obscure the code and they mean that the functions do more than one thing (which violates the single responsibility principle). It would be much cleaner to have two separate functions for the purpose. One calling another:

```c++
item &map::add_item_or_charges(const int x, const int y, const item &obj) {
  ...
}

item &map::add_item_or_charges_nearby(const int x, const int y, const item &obj) {
    item &res = add_item_or_charges( x, y, obj );

    if( !res.is_null() ) {
        return res;
    }

     // Do the overflow thing trying add_item_or_charges() on adjacent tiles
    ...
}
```

This would also prevent duplication in `if( valid_tile( e ) && !has_flag( "NOITEM", pos ) && valid_limits( e ) ) {`.

**Comment by mugling** [comment]
Well presuming we later drop the two potentially spurious uses of `overflow = false` then an entirely new function just for a single call isn't appealing either.

A documented function parameter is a reasonable design.

**Comment by codemime** [comment]
> Well presuming we later drop the two potentially spurious uses of overflow = false

You've already removed at least two usages of `2` from `item::process_litcig()` which changes the behaviour of the function (apart from the refactoring) and haven't provided a good reason for the change. You could have a pure refactoring using another function.

>  then an entirely new function just for a single call isn't appealing either.

I guess dropping an isolated function entirely later on is much easier than uprooting the flag from many places and reformatting the code afterwards.

**Comment by mugling** [comment]
> I guess dropping an isolated function entirely later on is much easier than uprooting the flag from many places and reformatting the code afterwards.

Silly pedantry. It's trivial to drop the argument from the function and then deal with the one solitary compilation error.

> Actually, considering that the argument 2 was removed from many functions (which likely has certain conesquences for the gameplay)

Likely? What? Can you provide a non-contrived argument to support that?

**Comment by remyroy** [comment]
The goal of this PR is good. I'm not sure how I can contribute to the discussion. In the maple code, the item should not overflow mainly because there shouldn't be any other item on the tapped maple tree. In any case, it would not make a huge difference if it was to overflow.

File: src/map.cpp [hunk 1]
```diff
@@ -4476,50 +4476,73 @@ units::volume map::free_volume(const tripoint &p) {
...
+    if( !valid_tile( pos ) ) {
+        return nulitem;
+    }
+
+    // If tile can contain items place here...
+    if( !has_flag( "NOITEM", pos ) && valid_limits( pos ) ) {
+        return place_item( pos );
+
+    // ...otherwise try to overflow to adjacent tiles (if permitted)
+    } else if( overflow ) {
```
Comment (Coolthulhu) [inline src/map.cpp]: Nit: The comment makes it look like the section above is in context of the function and not function+if.

**Comment by codemime** [comment]
> Silly pedantry. It's trivial to drop the argument from the function and then deal with the one solitary compilation error.
> A documented function parameter is a reasonable design versus multiple near-identical functions.

It's a code smell that should be avoided if possible. I guess it's tolerable in case of intermediate changes, 
but it certainly has nothing to do with a reasonable design.

> Likely? What? Can you provide a non-contrived argument to support that?

Well, the argument is a bit obvious: the radius was `2` and now it becomes `1`. This makes sudden item disappearance in an unchecked invocation of the function (vast majority of cases) three times more probable than it was before. This looks like a regression in a PR that's supposed to be an improvement and called "refactoring". If you really want to make things better, you could make the said unchecked disappearance simply impossible.

**Comment by Coolthulhu** [comment]
Fighting zombies in a corridor.
In this case it is possible for 3 tiles to overflow (a tile can have roughly 16 naked zed corpses, 10 fat zed corpses or 1 hulk).
It isn't a common case, but it can actually happen.

Picking 2 over 1 has the disadvantage of allowing items to fall through walls, but this is an incredibly rare case that pretty much requires intentional exploits.

Advanced inventory menu allows pilling objects over limit. This can trigger overflow.

**Comment by codemime** [comment]
> Picking 2 over 1 has the disadvantage of allowing items to fall through walls, but this is an incredibly rare case that pretty much requires intentional exploits.

May `map::clear_path()` be useful to prevent these rare cases? Although, it may be too restrictive if applied directly.

**Comment by mugling** [comment]
> Fighting zombies in a corridor.
> It isn't a common case, but it can actually happen.

Is this worth the extra complexity? The caller ought to be checking the return value if placement of the item is critical?

**Comment by Coolthulhu** [comment]
There isn't much extra complexity, just one number.

**Comment by mugling** [comment]
> There isn't much extra complexity, just one number.

The outcome is more complex though. Placed items can escape through walls and such things. Given the function is a primitive it can be confounded by other subtle bugs to produce further odd behaviors.

**Comment by Coolthulhu** [comment]
Yeah, but this is very rare and your alternative is those items totally disappearing.

**Comment by mugling** [comment]
Do you want radius 2 then?

**Comment by Coolthulhu** [comment]
Yeah, radius 2 would be safer. No one reported problems with radius 2.

**Comment by mugling** [comment]
> Yeah, radius 2 would be safer. No one reported problems with radius 2.

Done, ready

**Comment by mugling** [comment]
Rebased again

Ready

**Pull request merged**